### PR TITLE
[python3] Add vcpkg-cmake-wrapper.

### LINKED ISSUE #9026: [python3] cmake FindPython3 Development component
****

vcpkg does not present it's python3 port well to other ports and or there may be missing development files. cmake has built in support for detecting python3 and it is struggling to find the development component when python3 is installed via vcpkg.

For context, if you install python3 for Windows via the Visual Studio installer and a port you are building uses cmake and in the cmake files there is:

`find_package(Python3 COMPONENTS Development)`

https://cmake.org/cmake/help/v3.12/module/FindPython3.html

When the port is then built by vcpkg, cmake will pickup the python3 development component and the build will succeed.

If instead, you do not install python3 for Windows via the Visual Studio installer and simply add the dependency `python3` to the CONTROL file of the port and attempt to build, if you ensure that:

```
vcpkg_find_acquire_program(PYTHON3)
get_filename_component(PYTHON3_DIR "${PYTHON3}" DIRECTORY)
set(ENV{PATH} "$ENV{PATH};${PYTHON3_DIR}")
```

is inside the portfile, then the python exe is picked up but the development component is not found at all by cmake and the build will fail.

If you then go on to look at the python3 port's installed files and compare them to the python for Windows installed files, it appears that there may be a large number of files useful for developing python applications missing from vcpkg's python3 port's installation.

**Proposed solution**
`vcpkg_find_acquire_program(PYTHON3)` might benefit from special logic to determine if the python3 port has been built and installed so that it can present what cmake expects to find.

The expected development files are revealed to cmake via vcpkg's python3 port's installation or a python3-dev port is created to ensure that those who only want the lib get only what they need but the development files can be had separately.

**Describe alternatives you've considered**
I tried the MS Visual Studio Installer Python for Windows and it's system integration was much better than vcpkg's. cmake had absolutely no problems. Obviously, that kind of support would be nice in vcpkg.

**Additional context**
Python's pretty big now. It's become the common language at large and it's popularity continues to grow. [Python 2 has been officially announced as being depreciated ](https://github.com/python/devguide/pull/344) at the start of next year, so there will be lots of work going on now where developers are porting to Python 3.

Although vcpkg is primarily a C++ package manager, the growing number of C++ Python 3 bindings and the way in which vcpkg is especially useful for reproducible builds makes Python 3 integration on par with the integration found in the Visual Studio Installer an especially interesting goal.

Whilst it could be argued to just use Visual Studio's Python 3 installation, it appears more suitable to me, to use the Python 3 port instead.

### PR DESCRIPTION
**Pull request by Hoikas**
This adds a vcpkg-cmake-wrapper.cmake to the python3 port. CMake offers three disparate find modules that are compatible with Python 3.x. They are FindPythonInterp, FindPythonLibs, and FindPython. The first two are deprecated since CMake 3.12 but still widely used, and the last one implements FindPython2 and FindPython3. Working with these find modules is particularly challenging due to their propensity to find the Python installed by the system package manager or an installed Windows Python. FindPython particularly has issues resolving the Python libraries due to the separate debug/lib and lib directories in the vcpkg installed tree. This ensures that any calls to find_package() for PythonLibs, PythonInterp, and Python3 correctly finds the vcpkg provided python3 port. To see the improvement, look at the simplification to the python feature of itk.

I'd appreciate review, especially around the macOS portion of the wrapper. I don't have access to a mac machine, and the workarounds for static linkage are very empirical from my custom tests on GHA.

- What does your PR fix?
Fixes #9026
Closes #13556

- Which triplets are supported/not supported? Have you updated the CI baseline?
No change

- Does your PR follow the maintainer guide?
Yes

### CODE REVIEW

File: ports/python3/vcpkg-cmake-wrapper.cmake.in [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        find_package(Threads)
+        find_library(UTIL_LIBRARY NAMES util)
+        if(@PythonFinder_PREFIX@_LIBRARIES)
+            list(APPEND @PythonFinder_PREFIX@_LIBRARIES ${CMAKE_THREAD_LIBS_INIT} ${CMAKE_DL_LIBS} ${UTIL_LIBRARY})
+        endif()
+    endif()
+
+    if(APPLE)
+        find_package(Iconv)
+        find_package(Intl)
```
Comment (JackBoosY) [inline ports/python3/vcpkg-cmake-wrapper.cmake.in:82]:
```suggestion
        find_package(Iconv REQUIRED)
        find_package(Intl REQUIRED)
```

File: ports/python3/vcpkg-cmake-wrapper.cmake.in [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    unset(_PYTHON_LIB_DEBUG_VARNAME)
+endif()
+
+_find_package(${ARGS})
+
+if(@PythonFinder_LIBS@ AND @VCPKG_LIBRARY_LINKAGE@ STREQUAL static)
+    # Python for Windows embeds the zlib module into the core, so we have to link against it.
+    # This is a separate extension module on Unix-like platforms.
+    if(WIN32)
+        find_package(ZLIB)
```
Comment (JackBoosY) [inline ports/python3/vcpkg-cmake-wrapper.cmake.in:62]:
```suggestion
        find_package(ZLIB REQUIRED)
```

File: ports/python3/vcpkg-cmake-wrapper.cmake.in [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            list(APPEND @PythonFinder_PREFIX@_LIBRARIES ${ZLIB_LIBRARIES})
+        endif()
+        if(TARGET @PythonFinder_PREFIX@::Python)
+            set_property(TARGET @PythonFinder_PREFIX@::Python APPEND PROPERTY INTERFACE_LINK_LIBRARIES ZLIB::ZLIB)
+        endif()
+        if(TARGET @PythonFinder_PREFIX@::Module)
+            set_property(TARGET @PythonFinder_PREFIX@::Module APPEND PROPERTY INTERFACE_LINK_LIBRARIES ZLIB::ZLIB)
+        endif()
+    else()
+        find_package(Threads)
```
Comment (JackBoosY) [inline ports/python3/vcpkg-cmake-wrapper.cmake.in:73]:
```suggestion
        find_package(Threads REQUIRED)
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by BillyONeal** [comment]
FindPython is a builtin cmake module so this meets the usual "vcpkg-cmake-wrapper" nuclear option criteria. I do wonder though what our actual story is; we don't intend users to use the python port to get a `python.exe` or `pythonw.exe` or similar; that port is really about writing things that expose bindings to python rather than trying to be a complete python distribution channel. (i.e. we do not want to compete with Python's own MSIs or similar)

That said, the existing FindPython bits don't distinguish these use cases so we might be in a situation where we provide that accidentially; what effect does this have on people who use vcpkg but are expecting to find a system python where there might be pip packages or similar?

**Comment by Hoikas** [comment]
My focus in this PR is embedding Python into another application (in this case a 17 year old game). I agree that we don't want to replace the official Python distribution, but the separation of concerns as of Python 3 is very muddy. Python 3 is interesting in that the core actually imports Python code from its own standard library during initialization. If that code is unavailable, that is a fatal error.

In order to properly make the Python standard library available, FindPython can report the location of the standard library on the build machine for bundling but only if the executable is available. In the project I help maintain, we have to use `find_package(Python3 Interpreter Development REQUIRED)`, causing an awkward situation where a system Python with a minor version level matching the vcpkg port needed to be available for the interpreter component. #14891 helps with this slightly in that a matching interpreter is now available. But all of these finders by default go plundering in the registry and are incapable of finding the combination of the vcpkg interpreter and libraries without manual intervention**. FindPython tends to be very nondeterministic in error cases, returning either arbitrary system Python or failing entirely.

In #14891, I strove to make the vcpkg python3 "tool" look and behave similar to a system install of Python. When the port is built, we ensure that pip is installed, and that pip functionality is available. In the aforementioned project, we also have build scripts that need extensions installed by pip. This works by effectively `${Python3_EXECUTABLE} -m pip install -r requirements.txt`

** The manual intervention required is to pass something like this to cmake:
`-DPython3_FIND_REGISTRY=NEVER -DPython3_EXECUTABLE=<path to exe> -D_Python3_LIBRARY_RELEASE=<path to release lib> -D_Python3_LIBRARY_DEBUG=<path to debug lib>`
IMO this is burdensome to users of vcpkg and is the crux of the complaint in #9026.

**Comment by marekr** [comment]
> FindPython is a builtin cmake module so this meets the usual "vcpkg-cmake-wrapper" nuclear option criteria. I do wonder though what our actual story is; we don't intend users to use the python port to get a `python.exe` or `pythonw.exe` or similar; that port is really about writing things that expose bindings to python rather than trying to be a complete python distribution channel. (i.e. we do not want to compete with Python's own MSIs or similar)


@BillyONeal I'll give you my bit on this. I work on KiCad. We embed python as part of our application. This means we have a fully functioning python interpreter DLL involved but not the python exes.
However to provide our user the full python experience, we need to compile and include the python modules which currently are missing from vcpkg. Without these packages, even the python library being in vcpkg is rather completely worthless as a chunk of the expected core functionality is in them.

The python upstream distributions cannot be used for this as compiler settings have to match entirely across the whole build. In many cases there are libraries used by these core modules that vcpkg builds as well. This would mean we would need to somehow handle two separate copies of openssl at different versions for example.

**Comment by JackBoosY** [comment]
Ping @BillyONeal for response.

**Comment by BillyONeal** [comment]
I tagged with "requires: discussion" because I want it to be discussed; normally we discuss PRs with that tag on Thursdays but that didn't happen last week due to holidays :)

**Comment by BillyONeal** [comment]
OK, after some discussion it seems like the behavior should be:

* `find_package(python COMPONENTS interpreter)` should do nothing and just pass through to the built in find python
* `find_package(python COMPONENTS development)` should be overridden like you're trying to do here
* `find_package(python COMPONENTS interpreter development)` should be overridden even though the resulting interpreter might not be what the user wants

Was that your intent?

**Comment by Hoikas** [comment]
I think that would be a good match for what users expect and would satisfy my requirements. My only question is around the "old" pre-cmake 3.12 packages PythonLibs and PythonInterp. Those are separate calls, so it's hard to tell what the intention of the caller is. In this attempt, I attempted to handle everything in one find module. Maybe we should just... not care about the deprecated modules? I'm open to input on that.

**Comment by Hoikas** [comment]
Ok, I've written up the suggestions by @BillyONeal, and, as a bonus, added a `usage` file to point specifically to the CMake 3.12+ python finders. To recap, this is the expected behavior...

Per the [documentation](https://cmake.org/cmake/help/latest/module/FindPython.html):
```
If no COMPONENTS are specified, Interpreter is assumed.

If component Development is specified, it implies sub-components Development.Module and Development.Embed.
```

So:

`find_package(Python)` - implies `COMPONENTS Interpreter`, see below
`find_package(Python COMPONENTS Interpreter)` - does nothing
`find_package(Python COMPONENTS Development)` - implies `COMPONENTS Development.Embed Development.Module`, see below
`find_package(Python COMPONENTS Development.Embed)` - overrides both the interpreter and libraries
`find_package(Python COMPONENTS Development.Module)` - overrides only libraries
`find_package(Python COMPONENTS Interpreter Development.Module)` - overrides both the interpreter and libraries
`find_package(Python COMPONENTS Interpreter Development.Embed)` - overrides both the interpreter and libraries
`find_package(PythonInterp)` - does nothing
`find_package(PythonLibs)` - does nothing

Note that `Python3` is an alias for the `Python` package.

`Development.Embed` must override the interpreter so that the `Python_STDLIB` can be fetched and packaged as part of the embedder's build process.

**Comment by Hoikas** [comment]
I've pushed an important change that should prevent the python3 port from hiding other python interpreters on the path. This change should address CI failures seen in #15200 and #15298.

**Review by vicroms — APPROVED** [review]

**Comment by JackBoosY** [comment]
Note that this PR does not solve the issue that some ports rely on python2 instead of python3:
when python2 and python3 are downloaded to _downloads/tools/python_ at the same time, cppcms will always find python3 first, however, its script is version 2.x.

**Comment by Hoikas** [comment]
IIRC, cppcms does something strange with the `CMAKE_PROGRAM_PATH` instead of adding Python2 to the path. The next version of cppcms will support Python3 though :smile:

**Comment by JackBoosY** [comment]
@Hoikas Can you please fix it for the current version?

**Pull request merged**